### ISSUE TITLE
Category Derived columns

### ISSUE THREAD

**Issue by janagombitova** [akvo/akvo-lumen#2071]
# Context

**Why do we add this issue?** 

When researching the value of connecting Flow and Lumen, one of the questions we focused on understanding was Why use Lumen in the 1st place (and why do you not use Lumen)? In our research, we learned that Lumen's value is in making it simple to handle data. For complex tasks, you need to have a certain level of data skills and in that case, you usually have the tools you use already. But the use of data is spreading across development projects with the skills set of project officers not catching up yet. Plus working with data is an extra task on top of already a long to do list. So Lumen makes handling data easy to do and thus also saves time. 

In our research, we however also learned that our current derived column implementation does not meet the criteria of simplicity and some commonly needed and used data tasks are not yet supported with the other transformation options.  

# Problem or idea

**What is the status quo?**
Data comes in and often when visualising it Salim sees that there are too many variables making the visualisation hard to read. For example, when testing WQ the result numbers can vary but often ranges of these results are enough to indicate if the water is safe or not. Or another example would crop farmers plant. Out of the 10 different crops answered only 4 are most used and the other 6 can be grouped into one category called other and naming the answers, but keeping the visualisation with only 4 values. 

In order to make such changes to simplify the dataset and in consequence the visualisations (making them easier to read and understand) today Salim has to be familiar with javascript to use Lumen's derived column functionality. Salim, however, is not technical and does not know javascript making this either a task Akvo does for him what results in Salim not using Lumen or in Salim not using Lumen at all. 

**If this change is successful what might we observe?** 
* More partner users picking up using Lumen
* Less readership and requests about javascript Derived column
* Decrease in use of Derived column and increase in the new transformation
* Increase in visualisations 
* Increase in Lumen sales 

# Solution or next step

**How could we solve it?**
The idea is to make it easy to define new categories for existing values per column. 

**Who's behaviour will change and How?** 
*Salim*
We assume he will use Lumen more, spend more time with the dataset and transform it. We also assume he will create more visualisations (need for sharing best practices can be handy). 

We also assume we get more questions on where to do data cleaning, as the categorical column can be used for cluster-edit cases. What on one hand will prove the value of this change, but also the need to resolve the data questions across F and L. 

*Petra*
We assume Petra uses more complex formulas to track her M&E indicators thus the assumption is that this change will not be enough for her. We assume to hear more requests for expanding the functionality to fit her needs. Thus the categorical column transformation will be seen as useful, but not enough.

*Internal Akvo sales team* 
We assume their reluctance in selling Lumen and training on the use of Lumen will drop as the javascript barrier will be removed for the most common use case, meaning they do not have to explain what it is or how to use it, just state that it provides endless options for advanced users where the most common use cases are handled with the user friendly transformations.

**How will the new behaviours benefit users?**
* It will take them less time to visualise their data and gain insights
* It will give them insights on the survey structure and what to improve next time 

**How will the new behaviours benefit Akvo?** 
* We can be more comfortable selling Lumen as it fits our target users more
* Increase in income 
* Possible decrease in tech support
* Decrease in time spent creating partner dashboards --> can focus more on the more advanced analysis rather than doing to simple stuff partners can do themselves

**How can we measure if this is actually coming true?**
(see indicators tracked here: https://docs.google.com/spreadsheets/d/1H9PSqjG8xSnhgvzmOqztHjiaV9Ee4Erwm2FB_ZU9t8Q/edit#gid=1249264890)
* Decrease in use of derive column 
* Increase in use of new category column
* Increase in new Lumen contracts 
* Increase in new Lumen users
* Decrease in reads of the Cookbook

**Comment by janagombitova** [comment]
Here are the latest designs: https://projects.invisionapp.com/share/8CRV05F9TZW#/screens

A few things we decided for how the new Category column transformation should work: 
- if a value is given a category we by default give it category 'Uncategorised'. This way Salim can always see when new data comes in that he did not consider when creating the categorisation rule or when he did not handle all the data values.
- grouping values will not be restricted by the position of the value (meaning you can only group values on this row and under), but you can create a group of values from any value that has not been added to a group
- we should also add the option to search for values in the list and when creating a group 
- the UI should be styled in a way to clearly indicate that values are part of a group and that the value will be that category in the new column 
- the end categories will be of the data type TEXT. This needs to be visually shown in the UI
- we should add tracking to the new transformation option to see how the user adoption goes (did not mention in the meeting, forgot)

An extra side note: We see the value of being able to edit the categorisation rule for an existing column. But it can happen that other transformations depend on this column. To ensure later transformations do not break with editing the category rules we decided to set the data type of the new category column to TEXT. Editing the data type will not be possible, to make sure editing the rules is possible and does not break transformations. Editing a derived column from the transformation log will be part of the 2nd iteration and handled in a separate issue

**Comment by finnfiddle** [comment]
I need spanish and french translations for the following please:

@tangrammer @loicsans 

1. Source column: {title}
2. Derived Column Title
3. {number} unique values
4. {number} categories
5. Group
6. Uncategorized Values
7. Uncategorized
8. Select values to group
9. Select source column
10. Text
11. Number

i might have some of these already but Im sure it will take you a few extra seconds

**Comment by loicsans** [comment]
Source column: {title}: Colonne source: {titre}
Derived Column Title: Titre de colonne dérivé
{number} unique values: {nombre} valeurs uniques
{number} categories: {numéro} catégories
Group: Groupe 
Uncategorized Values: Valeurs non classées
Uncategorized: Non classé
Select values to group: Sélectionnez les valeurs à regrouper
Select source column: Sélectionner la colonne source
Text: Texte
Number: Nombre

**Comment by tangrammer** [comment]
Source column: {title}:  Columna de origen: {title}
Derived Column Title:  Título de columna derivada 
{number} unique values : valores únicos
{number} categories: categorías
Group: grupo
Uncategorized Values: Valores no categorizados
Uncategorized: Sin categorizar
Select values to group: Seleccionar valores a agrupar
Select source column: Seleccionar columna de origen
Text: texto 
Number: número

**Comment by finnfiddle** [comment]
I need a few more please:
`1 Category`, `X categories`
`1 Unique value`, `X unique values`
`Derive Column`
`New Category Column`
`Category Column`
`Via JavaScript`
Thanks in advance @loicsans @tangrammer

**Comment by loicsans** [comment]
1 Category, X categories: `1 catégorie`,` X catégories`
1 Unique value, X unique values: `1 valeur unique`,` X valeurs uniques`
Derive Column: `Dériver la colonne`
New Category Column: `Nouvelle Colonne de catégorie
Category Column: ` Colonne de catégorie`
Via JavaScript: `Via JavaScript`

**Comment by Kiarii** [comment]
category column:

# the column ‘pop-up’ how would it look if there lots of cols to select from
# Could we default the cursor (after the col selection) to the derive col name field?
# The alphabetical ordering (filter) may not apply for numbers, do we want to order the NUM values by their size and by the size of their count, or no sorting?
comments and thoughts..
# The group i suspect is still a WIP 
# the “uncategorized” values should count as a category, in which case, the system should show 5, not 4 categories..
# do we want to enforce category values and derived column name in a way that if the user attempts to [Derive the column] before filling all these in, there is a pointer, “red underline” for instance, just so the user can be aware of what the system requires of them?
<strikeout># Would be nice to make [Derive Column] button a callout button, so its easy to notice#</strikeout>
# When transformation is applied, ideally, the user should be moved back to the dataset, since there is not much to do in the current page.. or?

**Comment by finnfiddle** [comment]
@Kiarii 

> the column ‘pop-up’ how would it look if there lots of cols to select from

it would stretch down the page

> The alphabetical ordering (filter) may not apply for numbers, do we want to order the NUM values by their size and by the size of their count, or no sorting?

probably by their size makes sense

> The group i suspect is still a WIP

Is it that bad? ;) suggestions are welcome

> do we want to enforce category values and derived column name in a way that if the user attempts to [Derive the column] before filling all these in, there is a pointer, “red underline” for instance, just so the user can be aware of what the system requires of them?

I dont think we're able to guess which values he wants to assign categories to and which will fall under uncategorized. Or maybe i didn't understand correctly.

> Would be nice to make [Derive Column] button a callout button, so its easy to notice this need not be if the case above is handled clearly

Not sure what a callout button is but the button is the same as the other transformation types - ie merge datasets.

**Comment by janagombitova** [comment]
@finnfiddle and @Kiarii I gave it a go and here are some on my notes. Can you please connect on how to improve some of the following points? Thanks  

## Number columns should behave differently than text 
In the current implementation, there is no difference between how a text and a number column behave when creating a new category column. But in the designs, a number column is based on creating ranges rather than showing unique values and grouping them. 
 
<img width="1268" alt="Screen Shot 2019-05-21 at 09 14 43" src="https://user-images.githubusercontent.com/12456965/58075836-f9f7ce80-7ba8-11e9-923e-434dec31aec9.png">

## Should be visible that you can click on the Source column
You can change the source column, what is great, but it is not visible in first sight that this is possible. @Kiarii and @finnfiddle can you please connect on how to resolve this? 

![category selecting other column](https://user-images.githubusercontent.com/12456965/58076049-7ee2e800-7ba9-11e9-8f0d-dc3838af21b8.gif)

## Initial dropdown 
When selecting a transformation the category column is hidden under derived and only by hovering over you see there is more there. Can we visually indicate that the derive element in the list has more to it? 

<img width="1279" alt="Screen Shot 2019-05-21 at 09 26 17" src="https://user-images.githubusercontent.com/12456965/58076554-cc138980-7baa-11e9-8497-fc2099c44ec2.png">

## Groups need love 
I agree with Kiarii that how the groups look today is not ideal. They work well but the box is quite small and it just looks odd. I trust you two can figure out something better. 
<img width="1280" alt="Screen Shot 2019-05-21 at 09 32 11" src="https://user-images.githubusercontent.com/12456965/58076764-578d1a80-7bab-11e9-8c00-ef96974dfb28.png">

**Comment by Kiarii** [comment]
Thanks Jana for the point outs! 
The transformation will use an arrow to indicate extra options as follows:

![image](https://user-images.githubusercontent.com/12692444/58083174-2ae3fd80-7bc1-11e9-9363-4d362fb1a925.png)

**Comment by Kiarii** [comment]
Lets style the derive column input fields with the current styling for fields in lumen i.e. hide the field border and reveal it on hover, and include the active state.. this will lessen the visual noise.. 
![lumen-input field style](https://user-images.githubusercontent.com/12692444/58086214-e0fe1600-7bc6-11e9-9236-664aa329fa4a.gif)

**Comment by finnfiddle** [comment]
Iteration 1:

1. [ ] Lets style the derive column input fields with the current styling for fields in lumen i.e. hide the field border and reveal it on hover, and include the active state.. this will lessen the visual noise..
2. [ ] The transformation will use an arrow to indicate extra options as follows:
3. [ ] When I create categories we expect each to be unique. We should not allow Salim to create two identical categories. This should be validated at input.
4. [ ] groups layout
5. [ ] groups select all
6. [ ] groups close dialog
7. [ ] Should be visible that you can click on the Source column
8. [ ] Number columns should behave differently than text
9. [ ] the italics on "x unique values & x categories" look weird, could we remove the italic styling and user a smaller font for these labels?
10. [ ] Could we default the cursor (after the col selection) to the derive col name field?
11. [ ] The alphabetical ordering (filter) may not apply for numbers, do we want to order the NUM values by their size and by the size of their count, or no sorting?
comments and thoughts..
12. [ ] the “uncategorized” values should count as a category, in which case, the system should show 5, not 4 categories..
13. [ ] When transformation is applied, ideally, the user should be moved back to the dataset, since there is not much to do in the current page.. or?

**Comment by janagombitova** [comment]
@finnfiddle I checked the last box, as Juan resolved that this morning

**Comment by tangrammer** [comment]
Added a commit in BE in relation to "validate input" https://github.com/akvo/akvo-lumen/issues/2071#issuecomment-494274875

**Comment by tangrammer** [comment]
iteration 2:
- [ ] adapt client transformation log

**Comment by tangrammer** [comment]
backend side ready to play with column number type :) 
https://github.com/akvo/akvo-lumen/pull/2149
see PR description about payload details
```clojure
{:source {:column {:columnName "c3"}},
 :target {:column {:title "Column 1"}},
 :derivation
 {:uncategorizedValue "8",
  :mappings
  [[["<=" -1] ["<" -1] "0"]
   [[">" -1] ["=" -2.0] "q"]
   [[">=" -1] [">" 0] "0a"]
   [["<=" -1.0] ["<" 2.0] "Z5"]
   [["<" 1.0] [">=" 0] "G"]
   [[">" 0] [">" 2.0] "3"]
   [[">" -1] ["=" -1.0] "C"]
   [["<=" -1] ["=" -1] "1s"]
   [["=" 0] ["<" -1.0] "S"]
   [[">=" 1.0] nil "2tk1"]],
  :type "number"}}
```

**Comment by janagombitova** [comment]
@kardan I am reopening this issue to follow up on small fixes that need attention before we can push this out. 

@finnfiddle found two small bugs so far when testing (see images below) 
* [ ] selecting another column does not fit in its box 
* [ ] selecting All when grouping does not work

<img width="1275" alt="Screen Shot 2019-05-28 at 09 36 25" src="https://user-images.githubusercontent.com/12456965/58460032-0510a880-812d-11e9-98ed-169bb708133e.png">
<img width="1271" alt="Screen Shot 2019-05-28 at 09 39 06" src="https://user-images.githubusercontent.com/12456965/58460036-0641d580-812d-11e9-9387-bf2391be2231.png">

**Comment by janagombitova** [comment]
@finnfiddle I have one question too: When sorting alphabetically we make a difference between values if they start with a capital letter or not (screen grab first and last entry). Ideally, this should not matter, right? 
<img width="236" alt="Screen Shot 2019-05-28 at 09 45 57" src="https://user-images.githubusercontent.com/12456965/58460671-684f0a80-812e-11e9-9265-27910d1adb16.png">


Picture the use case in the example screenshot. The respondents answered the same thing but have written it differently. With this transformation (even though not really meant for this work) I could create groups with the same-ish values and thus clean my data. 

<img width="1278" alt="Screen Shot 2019-05-28 at 09 51 44" src="https://user-images.githubusercontent.com/12456965/58460687-6dac5500-812e-11e9-8933-30c733d00573.png">

**Comment by janagombitova** [comment]
@finnfiddle I am getting a blank page when trying to pick a few specific columns. Can you check? 

This dataset: 
https://lumen.akvotest.org/dataset/5ce52260-1c7a-4683-9f12-519b22b25c18?show-category-transform=1

Column: Please specify the non-kinh group or Why are you not picking
![Screen Shot 2019-05-29 at 13 54 53](https://user-images.githubusercontent.com/12456965/58555305-6615aa80-8219-11e9-9284-510f89274035.png)

**Comment by tangrammer** [comment]
to measure core/derive vs core/derive-category user adaptation: 

https://metrics.akvo.org/d/0HQ17qiWz/transformation-monitoring?orgId=1&fullscreen&panelId=5
